ludwig/config_validation/checks.py

```diff
@@ -0,0 +1,260 @@
+            return
+
+        input_feature_names = {input_feature[NAME] for input_feature in config[INPUT_FEATURES]}
+        for entity in ["entity_1", "entity_2"]:
+            for feature_name in config[COMBINER][entity]:
+                if feature_name not in input_feature_names:
+                    raise ConfigValidationError(
+                        f"Feature {feature_name} in {entity} for the comparator combiner is not a valid "
+                        "input feature name."
+                    )
+
+
+@register_config_check("Class balancing is only available for datasets with a single output feature.")
+class CheckClassBalancePreprocessing(ConfigCheck):
+    @staticmethod
+    def check(config: ModelConfigDict) -> None:
+        if config[PREPROCESSING]["oversample_minority"] or config[PREPROCESSING]["undersample_majority"]:
+            if len(config[OUTPUT_FEATURES]) != 1:
+                raise ConfigValidationError(
+                    "Class balancing is only available for datasets with a single output feature."
+                )
+            if config[OUTPUT_FEATURES][0][TYPE] != BINARY:
+                raise ConfigValidationError("Class balancing is only supported for binary output features.")
+
+
+@register_config_check("Oversample minority and undersample majority are mutually exclusive.")
+class CheckSamplingExclusivity(ConfigCheck):
+    @staticmethod
+    def check(config: ModelConfigDict) -> None:
+        if config[PREPROCESSING]["oversample_minority"] and config[PREPROCESSING]["undersample_majority"]:
+            raise ConfigValidationError(
+                "Oversample minority and undersample majority are mutually exclusive. Specify only one method."
+            )
+
+
+@register_config_check("Checks that the specified validation metric exists.")
+class CheckValidationMetricExists(ConfigCheck):
+    @staticmethod
+    def check(config: ModelConfigDict) -> None:
+        validation_metric_name = config[TRAINER]["validation_metric"]
+
+        # Get all valid metrics.
+        feature_to_metric_names_map = get_feature_to_metric_names_map(config[OUTPUT_FEATURES])
+        all_valid_metrics = set()
+        for metric_names in feature_to_metric_names_map.values():
+            all_valid_metrics.update(metric_names)
+
+        if validation_metric_name not in all_valid_metrics:
+            raise ConfigValidationError(
+                f"User-specified trainer.validation_metric '{validation_metric_name}' is not valid. "
+                f"Available metrics are: {all_valid_metrics}"
+            )
+
+
+@register_config_check("Checks the validity of the splitter configuration.")
+class CheckSplitter(ConfigCheck):
+    @staticmethod
+    def check(config: ModelConfigDict) -> None:
+        from ludwig.data.split import get_splitter
+
+        splitter = get_splitter(**config[PREPROCESSING][SPLIT])
+        splitter.validate(config)
+
+
+@register_config_check("Checks the schema.")
+class CheckSchema(ConfigCheck):
```

Comment:
Hmm, jw - does this cause the schema to be validated twice?

Reply:
Instantiating a model config will run schema validation twice -- once at the beginning of the constructor to check required fields (the rest of the config class code was written assuming that required fields are all present), and once at the end when we iterate through the registry.

Reply:
Why do it the second time?

Reply:
Removed the second schema check.